lib/autocomplete-manager.coffee

```diff
@@ -1,183 +1,111 @@
-  #
-  # editor - {TextEditor}
-  constructor: (@editor) ->
-    @editorView = atom.views.getView(@editor)
-    @compositeDisposable = new CompositeDisposable
+  editor: null
+  editorView: null
+  buffer: null
+  suggestionList: null
+  editorSubscription: null
+  bufferSavedSubscription: null
+  bufferChangedSubscription: null
+  editorCursorMovedSubscription: null
+
+  constructor: ->
+    @subscriptions = new CompositeDisposable
     @emitter = new Emitter
 
-    @providers = []
-
-    return if @currentFileBlacklisted()
+    @scopes = {}
+    @provideApi()
 
-    @registerProvider new FuzzyProvider(@editor)
+    # Register Suggestion List Model and View
+    @subscriptions.add(atom.views.addViewProvider(SuggestionList, (model) =>
+      new SuggestionListElement().initialize(model)
+    ))
+    @suggestionList = new SuggestionList()
 
     @handleEvents()
-    @setCurrentBuffer @editor.getBuffer()
-
-    @compositeDisposable.add atom.workspace.observeActivePaneItem(@updateCurrentEditor)
-
-    @compositeDisposable.add atom.commands.add 'atom-text-editor',
-      "autocomplete-plus:activate": @runAutocompletion
-
-    @compositeDisposable.add atom.commands.add 'autocomplete-suggestion-list',
-      "autocomplete-plus:confirm": @confirmSelection,
-      "autocomplete-plus:select-next": @selectNext,
-      "autocomplete-plus:select-previous": @selectPrevious,
-      "autocomplete-plus:cancel": @cancel
-
-  addKeyboardInteraction: ->
-    @removeKeyboardInteraction()
-    keys =
-      "escape": "autocomplete-plus:cancel"
-
-    completionKey = atom.config.get("autocomplete-plus.confirmCompletion") || ''
-    navigationKey = atom.config.get("autocomplete-plus.navigateCompletions") || ''
-
-
-    keys['tab'] = "autocomplete-plus:confirm" if completionKey.indexOf('tab') > -1
-    keys['enter'] = "autocomplete-plus:confirm" if completionKey.indexOf('enter') > -1
-
-    if @items?.length > 1 and navigationKey == "up,down"
-      keys['up'] =  "autocomplete-plus:select-previous"
-      keys['down'] = "autocomplete-plus:select-next"
-    else
-      keys["ctrl-n"] = "autocomplete-plus:select-next"
-      keys["ctrl-p"] = "autocomplete-plus:select-previous"
-
-    @keymaps = atom.keymaps.add(
-      'AutocompleteManager',
-      'atom-text-editor:not(.mini) .autocomplete-plus': keys
-    )
-
-    @compositeDisposable.add @keymaps
-
-  removeKeyboardInteraction: ->
-    @keymaps?.dispose()
-    @compositeDisposable.remove(@keymaps)
+    @handleCommands()
+    @fuzzyProvider = new FuzzyProvider()
 
   updateCurrentEditor: (currentPaneItem) =>
-    @cancel() unless currentPaneItem == @editor
-
-  confirmSelection: =>
-    @emitter.emit 'do-confirm-selection'
-
-  onDoConfirmSelection: (cb) ->
-    @emitter.on 'do-confirm-selection', cb
+    return unless currentPaneItem?
+    return if currentPaneItem is @editor
 
-  selectNext: =>
-    @emitter.emit 'do-select-next'
+    # Stop listening to buffer events
+    @bufferSavedSubscription?.dispose()
```

Comment:
We do maintain a `CompositeDisposable` but we cannot use that for buffer / editor events, because we have a single `AutocompleteManager` which is subscribing / unsubscribing as pane items change.

Consequently we need to explicitly track those subscriptions - hence the code is the way it is.